src/network.rs

```diff
@@ -0,0 +1,358 @@
+/// A module for native grid network functionality, so we don't have to depend on libcurl
+/// The crate-public functions are facades – they're designed for interaction with libproj –
+/// delegating actual functionality to non-public versions, prefixed by an underscore.
+///
+/// **Note**: `error_string_max_size` is set to 128 by libproj.
+// TODO: build some length checks for the errors that are stuffed into it
+// This functionality based on https://github.com/OSGeo/PROJ/blob/master/src/networkfilemanager.cpp#L1675
+use proj_sys::{proj_context_set_network_callbacks, PJ_CONTEXT, PROJ_NETWORK_HANDLE};
+
+use reqwest::blocking::{Client, RequestBuilder, Response};
+use reqwest::Method;
+use std::ffi::CString;
+use std::os::raw::c_ulonglong;
+use std::ptr;
+
+use crate::proj::{ProjError, _string};
+use libc::c_char;
+use libc::c_void;
+use std::boxed::Box;
+use std::{thread, time};
+
+const CLIENT: &str = concat!("proj-rs/", env!("CARGO_PKG_VERSION"));
+const MAX_RETRIES: u8 = 8;
+// S3 sometimes sends these in place of actual client errors, so retry instead of erroring
+const RETRY_CODES: [u16; 4] = [429, 500, 502, 504];
+
+/// This struct is cast to `c_void`, then to `PROJ_NETWORK_HANDLE` so it can be passed around
+#[no_mangle]
+struct HandleData {
+    request: reqwest::blocking::RequestBuilder,
+    headers: reqwest::header::HeaderMap,
+    // this raw pointer is returned to libproj but never returned from libproj,
+    // so a copy of the pointer (raw pointers are Copy) is stored here, so it can be
+    // reconstituted and dropped in network_close.
+    // Note to future self: are you 100% sure that the pointer is never read again
+    // after network_close returns?
+    hptr: Option<*const c_char>,
+}
+
+impl HandleData {
+    fn new(
+        request: reqwest::blocking::RequestBuilder,
+        headers: reqwest::header::HeaderMap,
+        hptr: Option<*const c_char>,
+    ) -> Self {
+        Self {
+            request,
+            headers,
+            hptr,
+        }
+    }
+}
+
+/// Return an exponential wait time based on the number of retries
+///
+/// Example: a value of 8 allows up to 6400 ms of retry delay, for a cumulative total of 25500 ms
+fn get_wait_time_exp(retrycount: i32) -> u64 {
+    if retrycount == 0 {
+        return 0;
+    }
+    (retrycount as u64).pow(2) * 100u64
+}
+
+/// Process CDN response: handle retries in case of server error, or early return for client errors
+fn error_handler<'a>(
+    res: &'a mut Response,
+    rb: RequestBuilder,
+    hvalue: &str,
+) -> Result<&'a Response, ProjError> {
+    let mut status = res.status().as_u16();
+    let mut retries = 0;
+    // Check whether something went wrong on the server, or if it's an S3 retry code
+    if res.status().is_server_error() || RETRY_CODES.contains(&status) {
+        // Start retrying: up to MAX_RETRIES
+        while res.status().is_server_error()
```

Comment:
So we'll retry infinitely if we keep getting server errors? 

My instinct would be to count server errors like other failures and only retry a limited number of times.

Something like `(res.status().is_server_error() || RETRY_CODES.contains(&status)) && retries <= MAX_RETRIES`

Reply:
I've just altered the retry condition on line 77  – we should exit the enclosing `if` on line 73 after 1 or `MAX_RETRIES` loops, failing on line 97 in case anything is still wrong.